File: src/baldr/laneconnectivity.cc
#include "baldr/laneconnectivity.h"
#include <boost/algorithm/string/classification.hpp>
#include <boost/algorithm/string/split.hpp>
#include <string.h>
#include <valhalla/midgard/logging.h>
#include <vector>

namespace valhalla {
namespace baldr {

// Constructor with arguments
LaneConnectivity::LaneConnectivity(const uint32_t idx,
                                   const uint64_t from,
                                   const std::string& to_lanes,
                                   const std::string& from_lanes)
    : to_(idx), from_(from), to_lanes_(to_lanes), from_lanes_(from_lanes) {
  if (from >= (1LL << 42)) {
    throw std::out_of_range("from way_id is too large");
  }
}

// Set the directed edge index.
void LaneConnectivity::set_to(const uint32_t idx) {
  to_ = idx;
}

// Returns the directed edge index.
uint32_t LaneConnectivity::to() const {
  return to_;
}

// Get the OSM id of the incoming way of this lane connection.
uint64_t LaneConnectivity::from() const {
  return from_;
}

// Get the text representation of lanes in the incoming way.
std::string LaneConnectivity::from_lanes() const {
  return from_lanes_.to_string();
}

// Get the text representation of lanes in the current way.
std::string LaneConnectivity::to_lanes() const {
  return to_lanes_.to_string();
}

// operator < - for sorting. Sort by `to` Id.
bool LaneConnectivity::operator<(const LaneConnectivity& other) const {
  return to() < other.to();
}

// Constructor with arguments.
LaneConnectivityLanes::LaneConnectivityLanes(const std::string& lanes) : value_(0) {
  std::vector<std::string> tokens;
  boost::split(tokens, lanes, boost::is_any_of("|"));
  uint8_t n = 1;
  for (const auto& t : tokens) {
    set_lane(n++, std::stoi(t));
  }
}

// Get the string representation of lane mask.
std::string LaneConnectivityLanes::to_string() const {
  std::string result;
  for (size_t i = 1; i <= kMaxLanesPerConnection; ++i) {
    uint8_t lane = get_lane(i);
    if (lane) {
      result += (result.empty() ? "" : "|") + std::to_string(lane);
    }
  }
  return result;
}

void LaneConnectivityLanes::set_lane(uint8_t n, uint8_t lane) {
  if (n == 0 || n > kMaxLanesPerConnection || lane > kMaxLanesPerConnection) {
    throw std::out_of_range("lane or index out of bounds");
  }
  value_ |= uint64_t(lane) << ((n - 1) * kMaxLanesPerConnectionBits);
}

uint8_t LaneConnectivityLanes::get_lane(uint8_t n) const {
  if (n == 0 || n > kMaxLanesPerConnection) {
    throw std::out_of_range("index out of bounds");
  }
  uint8_t v = (value_ >> ((n - 1) * kMaxLanesPerConnectionBits)) & kMaxLanesPerConnection;
  return v;
}

} // namespace baldr
} // namespace valhalla

File: src/baldr/laneconnectivity.cs
using System;
using System.Collections.Generic;
using System.Text.Json;

namespace valhalla.baldr
{
    public class LaneConnectivity
    {
        public uint to_ { get; set; }
        public ulong from_ { get; set; }
        public string to_lanes_ { get; set; }
        public string from_lanes_ { get; set; }

        public LaneConnectivity(uint idx, ulong from, string to_lanes, string from_lanes)
        {
            if (from >= (1UL << 42))
            {
                throw new ArgumentOutOfRangeException("from way_id is too large");
            }

            to_ = idx;
            from_ = from;
            to_lanes_ = to_lanes;
            from_lanes_ = from_lanes;
        }

        public void set_to(uint idx)
        {
            to_ = idx;
        }

        public uint to()
        {
            return to_;
        }

        public ulong from()
        {
            return from_;
        }

        public string from_lanes()
        {
            return from_lanes_;
        }

        public string to_lanes()
        {
            return to_lanes_;
        }

        public bool operator <(LaneConnectivity other)
        {
            return to() < other.to();
        }

        public JsonElement json()
        {
            var map = new Dictionary<string, object>
            {
                { "to", to_ },
                { "from", from_ },
                { "to_lanes", to_lanes_ },
                { "from_lanes", from_lanes_ }
            };

            return JsonDocument.Parse(JsonSerializer.Serialize(map)).RootElement;
        }
    }

    public class LaneConnectivityLanes
    {
        private ulong value_;

        public LaneConnectivityLanes(string lanes)
        {
            value_ = 0;
            var tokens = lanes.Split('|');
            byte n = 1;
            foreach (var t in tokens)
            {
                set_lane(n++, byte.Parse(t));
            }
        }

        public string to_string()
        {
            var result = string.Empty;
            for (int i = 1; i <= 8; ++i)
            {
                var lane = get_lane((byte)i);
                if (lane != 0)
                {
                    result += (result == string.Empty ? "" : "|") + lane.ToString();
                }
            }
            return result;
        }

        public void set_lane(byte n, byte lane)
        {
            if (n == 0 || n > 8 || lane > 8)
            {
                throw new ArgumentOutOfRangeException("lane or index out of bounds");
            }
            value_ |= (ulong)lane << ((n - 1) * 8);
        }

        public byte get_lane(byte n)
        {
            if (n == 0 || n > 8)
            {
                throw new ArgumentOutOfRangeException("index out of bounds");
            }
            return (byte)((value_ >> ((n - 1) * 8)) & 0xFF);
        }
    }
}
